File: core/services/aios-voice/voice.c
/**
 * AI-OS Voice Service
 * Speech recognition and TTS in C using PocketSphinx and eSpeak.
 * 
 * Compile: gcc -o aios-voice voice.c -lpocketsphinx -lsphinxbase -lasound -lpthread
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <pthread.h>
#include <errno.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <alsa/asoundlib.h>

/* PocketSphinx for speech recognition */
#ifdef USE_POCKETSPHINX
#include <pocketsphinx.h>
#endif

#define AGENT_SOCKET "/run/aios/agent.sock"
#define SAMPLE_RATE 16000
#define CHANNELS 1
#define WAKE_WORD "hey ai"

/* ==================== Globals ==================== */

static volatile int g_running = 1;
static int g_voice_enabled = 1;
static char g_wake_word[64] = WAKE_WORD;

/* ==================== Text-to-Speech ==================== */

static void tts_speak(const char *text) {
    if (!text || !text[0]) return;
    
    printf("[VOICE] Speaking: %s\n", text);
    
    /* Use espeak for TTS */
    char cmd[2048];
    snprintf(cmd, sizeof(cmd), "espeak -s 150 \"%s\" 2>/dev/null", text);
    system(cmd);
}

/* ==================== Agent Communication ==================== */

static int send_to_agent(const char *text, char *response, size_t response_size) {
    int sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) return -1;
    
    struct sockaddr_un addr = {0};
    addr.sun_family = AF_UNIX;
    strcpy(addr.sun_path, AGENT_SOCKET);
    
    if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        close(sock);
        return -1;
    }
    
    /* Build message */
    char msg[2048];
    snprintf(msg, sizeof(msg), "{\"cmd\":\"chat\",\"text\":\"%s\"}", text);
    
    uint32_t len = htonl(strlen(msg));
    send(sock, &len, 4, 0);
    send(sock, msg, strlen(msg), 0);
    
    /* Receive response */
    if (recv(sock, &len, 4, 0) != 4) {
        close(sock);
        return -1;
    }
    len = ntohl(len);
    
    if (len < response_size) {
        recv(sock, response, len, 0);
        response[len] = '\0';
    }
    
    close(sock);
    return 0;
}

static void process_command(const char *command) {
    printf("[VOICE] Processing: %s\n", command);
    
    char response[4096] = {0};
    if (send_to_agent(command, response, sizeof(response)) == 0) {
        /* Extract response text */
        char *p = strstr(response, "\"response\":\"");
        if (p) {
            p += 12;
            char *end = strchr(p, '"');
            if (end) {
                *end = '\0';
                /* Unescape */
                char clean[1024];
                char *src = p, *dst = clean;
                while (*src && dst < clean + sizeof(clean) - 1) {
                    if (*src == '\\' && *(src+1) == 'n') {
                        *dst++ = ' ';
                        src += 2;
                    } else {
                        *dst++ = *src++;
                    }
                }
                *dst = '\0';
                tts_speak(clean);
            }
        }
    } else {
        tts_speak("Sorry, I couldn't connect to the agent.");
    }
}

/* ==================== Audio Capture ==================== */

static snd_pcm_t *open_audio_capture(void) {
    snd_pcm_t *handle;
    snd_pcm_hw_params_t *params;
    int err;
    
    err = snd_pcm_open(&handle, "default", SND_PCM_STREAM_CAPTURE, 0);
    if (err < 0) {
        fprintf(stderr, "[VOICE] Cannot open audio: %s\n", snd_strerror(err));
        return NULL;
    }
    
    snd_pcm_hw_params_alloca(&params);
    snd_pcm_hw_params_any(handle, params);
    snd_pcm_hw_params_set_access(handle, params, SND_PCM_ACCESS_RW_INTERLEAVED);
    snd_pcm_hw_params_set_format(handle, params, SND_PCM_FORMAT_S16_LE);
    snd_pcm_hw_params_set_channels(handle, params, CHANNELS);
    
    unsigned int rate = SAMPLE_RATE;
    snd_pcm_hw_params_set_rate_near(handle, params, &rate, 0);
    
    snd_pcm_hw_params_set_period_size(handle, params, 1024, 0);
    snd_pcm_hw_params_set_buffer_size(handle, params, 8192);
    
    err = snd_pcm_hw_params(handle, params);
    if (err < 0) {
        fprintf(stderr, "[VOICE] Cannot set params: %s\n", snd_strerror(err));
        snd_pcm_close(handle);
        return NULL;
    }
    
    return handle;
}

/* ==================== Simple Voice Activity Detection ==================== */

static int detect_voice_activity(int16_t *buffer, int frames) {
    int32_t sum = 0;
    for (int i = 0; i < frames; i++) {
        sum += abs(buffer[i]);
    }
    int avg = sum / frames;
    return avg > 500;  /* Threshold */
}

/* ==================== Speech Recognition Loop ==================== */

#ifdef USE_POCKETSPHINX

static void recognition_loop(void) {
    snd_pcm_t *audio = open_audio_capture();
    if (!audio) {
        fprintf(stderr, "[VOICE] No audio device\n");
        return;
    }
    
    /* Initialize PocketSphinx */
    ps_config_t *config = ps_config_init(NULL);
    ps_config_set_str(config, "hmm", "/usr/share/pocketsphinx/model/en-us/en-us");
    ps_config_set_str(config, "lm", "/usr/share/pocketsphinx/model/en-us/en-us.lm.bin");
    ps_config_set_str(config, "dict", "/usr/share/pocketsphinx/model/en-us/cmudict-en-us.dict");
    
    ps_decoder_t *decoder = ps_init(config);
    if (!decoder) {
        fprintf(stderr, "[VOICE] Failed to init decoder\n");
        snd_pcm_close(audio);
        return;
    }
    
    ps_start_utt(decoder);
    
    int16_t buffer[1024];
    int listening_for_command = 0;
    int silence_frames = 0;
    char command[256] = {0};
    
    printf("[VOICE] Listening for wake word: '%s'\n", g_wake_word);
    
    while (g_running) {
        int frames = snd_pcm_readi(audio, buffer, 1024);
        if (frames < 0) {
            snd_pcm_recover(audio, frames, 0);
            continue;
        }
        
        ps_process_raw(decoder, buffer, frames, FALSE, FALSE);
        
        const char *hyp = ps_get_hyp(decoder, NULL);
        
        if (hyp) {
            char lower[256];
            strncpy(lower, hyp, sizeof(lower) - 1);
            for (char *p = lower; *p; p++) *p = (*p >= 'A' && *p <= 'Z') ? *p + 32 : *p;
            
            if (!listening_for_command) {
                if (strstr(lower, g_wake_word)) {
                    printf("[VOICE] Wake word detected!\n");
                    tts_speak("Yes?");
                    listening_for_command = 1;
                    command[0] = '\0';
                    ps_end_utt(decoder);
                    ps_start_utt(decoder);
                }
            } else {
                /* Accumulate command */
                strncpy(command, hyp, sizeof(command) - 1);
                
                /* Check for silence */
                if (!detect_voice_activity(buffer, frames)) {
                    silence_frames++;
                    if (silence_frames > 20) {  /* ~1 second of silence */
                        if (command[0]) {
                            process_command(command);
                        }
                        listening_for_command = 0;
                        silence_frames = 0;
                        ps_end_utt(decoder);
                        ps_start_utt(decoder);
                    }
                } else {
                    silence_frames = 0;
                }
            }
        }
    }
    
    ps_end_utt(decoder);
    ps_free(decoder);
    ps_config_free(config);
    snd_pcm_close(audio);
}

#else

/* Simplified loop without PocketSphinx - uses external speech-to-text */
static void recognition_loop(void) {
    printf("[VOICE] Running in voice activity detection mode\n");
    printf("[VOICE] Say '%s' to activate\n", g_wake_word);
    
    snd_pcm_t *audio = open_audio_capture();
    if (!audio) {
        /* Fallback to console input */
        printf("[VOICE] No audio, using console input\n");
        
        char line[256];
        while (g_running) {
            printf("Voice> ");
            if (!fgets(line, sizeof(line), stdin)) break;
            
            line[strcspn(line, "\n")] = '\0';
            if (line[0]) {
                process_command(line);
            }
        }
        return;
    }
    
    int16_t buffer[1024];
    int was_speaking = 0;
    int silence_count = 0;
    
    while (g_running) {
        int frames = snd_pcm_readi(audio, buffer, 1024);
        if (frames < 0) {
            snd_pcm_recover(audio, frames, 0);
            continue;
        }
        
        int is_speaking = detect_voice_activity(buffer, frames);
        
        if (is_speaking && !was_speaking) {
            printf("[VOICE] Voice activity detected\n");
        } else if (!is_speaking && was_speaking) {
            silence_count++;
            if (silence_count > 10) {
                printf("[VOICE] Voice ended\n");
                /* Would trigger STT here */
                silence_count = 0;
            }
        }
        
        was_speaking = is_speaking;
        usleep(10000);
    }
    
    snd_pcm_close(audio);
}

#endif

/* ==================== Main ==================== */

static void signal_handler(int sig) {
    g_running = 0;
}

int main(int argc, char *argv[]) {
    printf("[VOICE] AI-OS Voice Service starting...\n");
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    /* Load config */
    FILE *f = fopen("/etc/aios/voice.json", "r");
    if (f) {
        char buf[1024];
        fread(buf, 1, sizeof(buf), f);
        fclose(f);
        
        char *p = strstr(buf, "\"wake_word\":\"");
        if (p) {
            sscanf(p, "\"wake_word\":\"%63[^\"]\"", g_wake_word);
        }
        
        if (strstr(buf, "\"enabled\":false")) {
            g_voice_enabled = 0;
        }
    }
    
    if (!g_voice_enabled) {
        printf("[VOICE] Voice service disabled\n");
        while (g_running) sleep(60);
        return 0;
    }
    
    tts_speak("AI-OS voice service ready");
    
    recognition_loop();
    
    printf("[VOICE] Voice service stopped\n");
    return 0;
}

File: core/services/aios-voice/voice.py
#!/usr/bin/env python3
"""
AI-OS Voice Service
Handles wake word detection, speech recognition, and TTS.
"""

import os
import sys
import json
import socket
import struct
import signal
import logging
import threading
import time
from pathlib import Path
from typing import Optional
from dataclasses import dataclass

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(name)s: %(message)s',
    handlers=[
        logging.FileHandler("/var/log/aios/voice.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger('aios-voice')


@dataclass
class VoiceConfig:
    """Voice service configuration"""
    enabled: bool = True
    wake_word: str = "hey ai"
    stt_engine: str = "vosk"  # vosk, google
    tts_engine: str = "espeak"  # espeak, pyttsx3
    sample_rate: int = 16000
    vad_aggressiveness: int = 3  # 0-3
    vosk_model_path: str = "/usr/share/vosk-models/small-en-us"
    
    @classmethod
    def load(cls) -> 'VoiceConfig':
        config = cls()
        config_file = Path("/etc/aios/voice.json")
        if config_file.exists():
            try:
                with open(config_file) as f:
                    data = json.load(f)
                for key, value in data.items():
                    if hasattr(config, key):
                        setattr(config, key, value)
            except Exception as e:
                logger.warning(f"Failed to load config: {e}")
        return config


class AgentClient:
    """Client for agent daemon"""
    
    SOCKET_PATH = "/run/aios/agent.sock"
    
    def chat(self, text: str) -> str:
        try:
            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
                sock.connect(self.SOCKET_PATH)
                
                message = json.dumps({'cmd': 'chat', 'text': text})
                data = message.encode('utf-8')
                sock.sendall(struct.pack('!I', len(data)))
                sock.sendall(data)
                
                length_data = sock.recv(4)
                if not length_data:
                    return "No response from agent"
                
                length = struct.unpack('!I', length_data)[0]
                response_data = sock.recv(length)
                response = json.loads(response_data.decode('utf-8'))
                
                return response.get('response', 'No response')
        except Exception as e:
            return f"Error: {e}"


class TextToSpeech:
    """Text-to-Speech engine"""
    
    def __init__(self, engine: str = "espeak"):
        self.engine = engine
        self._pyttsx3 = None
        
        if engine == "pyttsx3":
            try:
                import pyttsx3
                self._pyttsx3 = pyttsx3.init()
                self._pyttsx3.setProperty('rate', 150)
            except ImportError:
                self.engine = "espeak"
    
    def speak(self, text: str):
        """Speak text"""
        logger.info(f"Speaking: {text[:50]}...")
        
        if self.engine == "pyttsx3" and self._pyttsx3:
            self._pyttsx3.say(text)
            self._pyttsx3.runAndWait()
        else:
            # Use espeak
            import subprocess
            try:
                subprocess.run(
                    ['espeak', '-s', '150', text],
                    check=True,
                    capture_output=True
                )
            except Exception as e:
                logger.error(f"TTS error: {e}")


class VoiceService:
    """Main voice service"""
    
    def __init__(self, config: VoiceConfig):
        self.config = config
        self.running = False
        self.agent = AgentClient()
        self.tts = TextToSpeech(config.tts_engine)
        
        self._recognizer = None
        self._microphone = None
        
    def start(self):
        """Start voice service"""
        if not self.config.enabled:
            logger.info("Voice service disabled")
            return
        
        logger.info("Starting AI-OS Voice Service...")
        
        # Initialize speech recognition
        if not self._init_recognition():
            logger.error("Failed to initialize speech recognition")
            return
        
        self.running = True
        
        # Announce startup
        self.tts.speak("AI-OS voice service ready. Say 'Hey AI' to activate.")
        
        # Start listening
        self._listen_loop()
    
    def _init_recognition(self) -> bool:
        """Initialize speech recognition"""
        try:
            import speech_recognition as sr
            self._recognizer = sr.Recognizer()
            self._microphone = sr.Microphone(sample_rate=self.config.sample_rate)
            
            # Adjust for ambient noise
            with self._microphone as source:
                logger.info("Calibrating for ambient noise...")
                self._recognizer.adjust_for_ambient_noise(source, duration=2)
            
            logger.info("Speech recognition initialized")
            return True
            
        except ImportError:
            logger.error("speech_recognition not available")
            return False
        except Exception as e:
            logger.error(f"Failed to init recognition: {e}")
            return False
    
    def _listen_loop(self):
        """Main listening loop"""
        import speech_recognition as sr
        
        logger.info("Listening for wake word...")
        
        while self.running:
            try:
                with self._microphone as source:
                    try:
                        audio = self._recognizer.listen(source, timeout=5)
                    except sr.WaitTimeoutError:
                        continue
                
                # Recognize speech
                text = self._recognize(audio)
                if not text:
                    continue
                
                text_lower = text.lower()
                logger.info(f"Heard: {text}")
                
                # Check for wake word
                if self.config.wake_word in text_lower:
                    # Extract command after wake word
                    idx = text_lower.find(self.config.wake_word)
                    command = text[idx + len(self.config.wake_word):].strip()
                    
                    if command:
                        self._process_command(command)
                    else:
                        # Wait for command
                        self.tts.speak("Yes?")
                        
                        with self._microphone as source:
                            try:
                                audio = self._recognizer.listen(source, timeout=5)
                                command = self._recognize(audio)
                                if command:
                                    self._process_command(command)
                            except sr.WaitTimeoutError:
                                self.tts.speak("I didn't hear anything.")
                
            except KeyboardInterrupt:
                break
            except Exception as e:
                logger.error(f"Listening error: {e}")
                time.sleep(1)
    
    def _recognize(self, audio) -> Optional[str]:
        """Recognize speech from audio"""
        import speech_recognition as sr
        
        try:
            if self.config.stt_engine == "vosk":
                # Try Vosk
                try:
                    return self._recognizer.recognize_vosk(
                        audio,
                        model_path=self.config.vosk_model_path
                    )
                except:
                    pass
            
            # Try Google
            return self._recognizer.recognize_google(audio)
            
        except sr.UnknownValueError:
            return None
        except sr.RequestError as e:
            logger.warning(f"Recognition service error: {e}")
            return None
        except Exception as e:
            logger.error(f"Recognition error: {e}")
            return None
    
    def _process_command(self, command: str):
        """Process voice command"""
        logger.info(f"Processing command: {command}")
        
        try:
            # Send to agent
            response = self.agent.chat(command)
            
            # Extract text response (remove JSON if present)
            import re
            response_text = re.sub(r'```json[\s\S]*?```', '', response).strip()
            
            if response_text:
                self.tts.speak(response_text)
            else:
                self.tts.speak("Done.")
                
        except Exception as e:
            logger.error(f"Command processing error: {e}")
            self.tts.speak("Sorry, I encountered an error.")
    
    def stop(self):
        """Stop voice service"""
        self.running = False
        logger.info("Voice service stopped")


def main():
    """Main entry point"""
    os.makedirs("/var/log/aios", exist_ok=True)
    
    config = VoiceConfig.load()
    service = VoiceService(config)
    
    def signal_handler(sig, frame):
        service.stop()
        sys.exit(0)
    
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    
    service.start()


if __name__ == '__main__':
    main()
